vnext/Microsoft.ReactNative/XamlLoadState.cpp

```diff
@@ -0,0 +1,178 @@
+  ULONG SizeOfImage; // The size of the DLL image, in bytes.
+} LDR_DLL_LOADED_NOTIFICATION_DATA, *PLDR_DLL_LOADED_NOTIFICATION_DATA;
+
+typedef struct _LDR_DLL_UNLOADED_NOTIFICATION_DATA {
+  ULONG Flags; // Reserved.
+  PCUNICODE_STRING FullDllName; // The full path name of the DLL module.
+  PCUNICODE_STRING BaseDllName; // The base file name of the DLL module.
+  PVOID DllBase; // A pointer to the base address for the DLL in memory.
+  ULONG SizeOfImage; // The size of the DLL image, in bytes.
+} LDR_DLL_UNLOADED_NOTIFICATION_DATA, *PLDR_DLL_UNLOADED_NOTIFICATION_DATA;
+
+typedef union _LDR_DLL_NOTIFICATION_DATA {
+  LDR_DLL_LOADED_NOTIFICATION_DATA Loaded;
+  LDR_DLL_UNLOADED_NOTIFICATION_DATA Unloaded;
+} LDR_DLL_NOTIFICATION_DATA, *PLDR_DLL_NOTIFICATION_DATA;
+
+typedef LDR_DLL_NOTIFICATION_DATA const *PCLDR_DLL_NOTIFICATION_DATA;
+
+typedef VOID CALLBACK LDR_DLL_NOTIFICATION_FUNCTION(
+    _In_ ULONG NotificationReason,
+    _In_ PCLDR_DLL_NOTIFICATION_DATA NotificationData,
+    _In_opt_ PVOID Context);
+
+typedef LDR_DLL_NOTIFICATION_FUNCTION *PLDR_DLL_NOTIFICATION_FUNCTION;
+
+#define LDR_DLL_NOTIFICATION_REASON_LOADED 1
+
+NTSTATUS NTAPI LdrRegisterDllNotification(
+    _In_ ULONG Flags,
+    _In_ PLDR_DLL_NOTIFICATION_FUNCTION NotificationFunction,
+    _In_opt_ PVOID Context,
+    _Out_ PVOID *Cookie);
+
+NTSTATUS NTAPI LdrUnregisterDllNotification(_In_ PVOID Cookie);
+
+#pragma endregion
+
+LDR_DLL_NOTIFICATION_FUNCTION MyNotificationFunction;
+
+#define NTDLL_FUNCTION(name) reinterpret_cast<decltype(&name)>(GetProcAddress(ntdll, #name))
+
+std::string XamlLoadState::GetXamlVersion(PCWSTR path) {
+  DWORD dwHandle;
+  DWORD size = GetFileVersionInfoSizeW(path, &dwHandle);
+  std::string buffer;
+  buffer.reserve(size);
+
+  if (GetFileVersionInfo(path, 0, size, const_cast<char *>(buffer.c_str()))) {
+    VS_FIXEDFILEINFO *fileInfo;
+    UINT len{};
+    if (VerQueryValueW(buffer.c_str(), L"\\", (void **)&fileInfo, &len)) {
+      const DWORD ls = fileInfo->dwFileVersionMS & 0x3fff;
+      DWORD ms = fileInfo->dwFileVersionMS >> 16;
+      // WinUI 3 Alpha has version 10.0.10011.16384. Convert 10.0 to 3.0
+      if (ms == 10) {
+        ms = 3;
+      }
+      std::stringstream v;
+      v << ms << "." << ls;
+      return v.str();
+    }
+  }
+
+  return {};
+}
+
+XamlLoadState::XamlLoadState() {
+  HMODULE mod{nullptr};
+
+  if (GetModuleHandle(wux.data())) {
+    RegisterDll(wux.data(), nullptr);
+  }
+  mod = GetModuleHandle(mux.data());
+  if (mod) {
+    wchar_t path[MAX_PATH]{};
+    if (GetModuleFileNameW(mod, path, ARRAYSIZE(path))) {
+      RegisterDll(mux.data(), path);
+    }
+  }
+
+  auto ntdll = GetModuleHandle(L"ntdll.dll");
+  auto pfn = NTDLL_FUNCTION(LdrRegisterDllNotification);
+  (*pfn)(0, MyNotificationFunction, nullptr, &m_cookie);
+}
+
+XamlLoadState::~XamlLoadState() {
+  auto ntdll = GetModuleHandle(L"ntdll.dll");
+  auto pfn = NTDLL_FUNCTION(LdrUnregisterDllNotification);
```

Comment:
Is it safe to assume we can still call this during static variable destruction?

Reply:
yes, ntdll gets unloaded much later than when static variable destructors